internal/controller/ngrok/cloudendpoint_controller.go

```diff
@@ -0,0 +1,342 @@
+		Create:   r.create,
+		Update:   r.update,
+		Delete:   r.delete,
+	}
+
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &ngrokv1alpha1.CloudEndpoint{}, trafficPolicyNameIndex, func(o client.Object) []string {
+		clep, ok := o.(*ngrokv1alpha1.CloudEndpoint)
+		if !ok {
+			return nil
+		}
+		return []string{clep.Spec.TrafficPolicyName}
+	}); err != nil {
+		return err
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&ngrokv1alpha1.CloudEndpoint{}).
+		Watches(
+			&ngrokv1alpha1.NgrokTrafficPolicy{},
+			handler.EnqueueRequestsFromMapFunc(r.findCloudEndpointForTrafficPolicy),
+			// Don't process delete events as it will just fail to look it up.
+			// Instead rely on the user to either delete the CloudEndpoint CR or update it with a new TrafficPolicy name
+			builder.WithPredicates(&predicate.Funcs{
+				DeleteFunc: func(e event.DeleteEvent) bool {
+					return false
+				},
+			}),
+		).
+		WithEventFilter(
+			predicate.Or(
+				predicate.GenerationChangedPredicate{},
+			),
+		).
+		Complete(r)
+}
+
+// Define a custom error type to signal a delayed requeue
+var ErrDomainCreating = errors.New("domain is being created, requeue after delay")
+
+// #region Reconcile CRUD
+
+// +kubebuilder:rbac:groups=ngrok.k8s.ngrok.com,resources=cloudendpoints,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=ngrok.k8s.ngrok.com,resources=cloudendpoints/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=ngrok.k8s.ngrok.com,resources=cloudendpoints/finalizers,verbs=update
+
+func (r *CloudEndpointReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	clep := new(ngrokv1alpha1.CloudEndpoint)
+
+	// Call the base controller's Reconcile method
+	result, err := r.controller.Reconcile(ctx, req, clep)
+
+	// Handle the custom error for domain creation delay
+	if errors.Is(err, ErrDomainCreating) {
+		// Requeue after 10 seconds to allow time for domain registration
+		return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+	}
+
+	return result, err
+}
+
+// Create will make sure a domain is created before creating the Cloud Endpoint
+// It also looks up the Traffic Policy and creates the Cloud Endpoint using this Traffic Policy JSON
+func (r *CloudEndpointReconciler) create(ctx context.Context, clep *ngrokv1alpha1.CloudEndpoint) error {
+	err := r.ensureDomainExists(ctx, clep)
+	if err != nil {
+		return err
+	}
+
+	policy, err := r.findTrafficPolicy(ctx, clep.Spec.TrafficPolicyName, clep.Namespace)
+	if err != nil {
+		return err
+	}
+
+	createParams := &ngrok.EndpointCreate{
+		Type:          "cloud",
+		URL:           clep.Spec.URL,
+		Description:   &clep.Spec.Description,
+		Metadata:      &clep.Spec.Metadata,
+		TrafficPolicy: policy,
+		Bindings:      clep.Spec.Bindings,
+	}
+
+	ngrokClep, err := r.NgrokClientset.Endpoints().Create(ctx, createParams)
+	if err != nil {
+		return err
+	}
+
+	return r.updateStatus(ctx, clep, ngrokClep)
```

Comment:
Note: I'm working on a new status pattern to interact with the base controller as well. This pattern will help handle errors, events, and logs for status updates.

Non-blocking: We can come back and add this in the future.